File: compiler/typecheck/instantiate_constraint.go
package typecheck

import (
	"wipple/database"
)

type Instantiation struct {
	Source        database.Node
	Definition    database.Node
	Replacements  map[database.Node]database.Node
	Substitutions *map[database.Node]Type
	KeepGeneric   bool // used during instance overlap checking
}

type InstantiateConstraint struct {
	info                     *ConstraintInfo
	Instantiation            Instantiation
	GetDefinitionConstraints func(database.Node) []Constraint
}

func (c *InstantiateConstraint) Info() *ConstraintInfo {
	return c.info
}

func (c *InstantiateConstraint) String() string {
	return "InstantiateConstraint(...)"
}

func (c *InstantiateConstraint) Instantiate(solver *Solver, source database.Node, replacements map[database.Node]database.Node, substitutions *map[database.Node]Type) Constraint {
	newReplacements := make(map[database.Node]database.Node, len(c.Instantiation.Replacements))
	for node, replacement := range c.Instantiation.Replacements {
		newReplacements[node] = GetOrInstantiate(solver, replacement, source, replacements)
	}

	newSubstitutions := make(map[database.Node]Type, len(*c.Instantiation.Substitutions))
	for node, substitution := range *c.Instantiation.Substitutions {
		newSubstitutions[node] = InstantiateType(solver, substitution, source, substitutions, replacements)
	}

	return NewInstantiateConstraint(Instantiation{
		Source:        source,
		Definition:    c.Instantiation.Definition,
		Replacements:  newReplacements,
		Substitutions: &newSubstitutions,
	}, c.GetDefinitionConstraints)
}

func (c *InstantiateConstraint) Run(solver *Solver) bool {
	// NOTE: Types are *not* applied before instantiating; we have access to
	// all related nodes/constraints here, which together will form better
	// groups

	definitionConstraints := c.GetDefinitionConstraints(c.Instantiation.Definition)

	constraints := make([]Constraint, 0, len(definitionConstraints))
	for _, constraint := range definitionConstraints {
		if !constraint.Info().ShouldInstantiate {
			continue
		}

		if !c.Instantiation.KeepGeneric {
			constraint = constraint.Instantiate(solver, c.Instantiation.Source, c.Instantiation.Replacements, c.Instantiation.Substitutions)
		}

		constraints = append(constraints, constraint)
	}

	solver.Constraints.Add(constraints...)

	return true
}

func NewInstantiateConstraint(i Instantiation, getDefinitionConstraints func(database.Node) []Constraint) Constraint {
	return &InstantiateConstraint{
		info:                     DefaultConstraintInfo(nil),
		Instantiation:            i,
		GetDefinitionConstraints: getDefinitionConstraints,
	}
}

File: compiler/core/src/typecheck/constraints/instantiate_constraint.rs
use crate::{
    db::{Db, Node},
    typecheck::{
        constraints::{
            AnyConstraintTrace, Constraint, ConstraintKind, ConstraintTrace, RunResult, Solver,
        },
        instantiate::InstantiateCtx,
        solver::SubstitutionsKey,
    },
    visit::DefinitionConstraints,
};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InstantiateConstraint {
    pub source_node: Node,
    pub bound_path: Vec<Node>,
    pub definition: Node,
    pub substitutions: SubstitutionsKey,
    pub traces: Vec<AnyConstraintTrace>,
}

impl InstantiateConstraint {
    pub fn new(source_node: Node, definition: Node, substitutions: SubstitutionsKey) -> Self {
        InstantiateConstraint {
            source_node,
            bound_path: Vec::new(),
            definition,
            substitutions,
            traces: Vec::new(),
        }
    }

    pub fn with_trace(mut self, trace: impl ConstraintTrace) -> Self {
        self.traces.push(AnyConstraintTrace::new(trace));
        self
    }
}

#[typetag::serde]
impl Constraint for InstantiateConstraint {
    fn kind(&self) -> ConstraintKind {
        ConstraintKind::Ty
    }

    fn node(&self) -> Node {
        self.source_node
    }

    fn traces_mut(&mut self) -> &mut Vec<AnyConstraintTrace> {
        &mut self.traces
    }

    fn instantiate(
        &self,
        db: &mut Db,
        solver: &mut Solver,
        ctx: &mut InstantiateCtx,
    ) -> Option<Box<dyn Constraint>> {
        let (mut nodes, mut parameters) = solver.get_substitutions(self.substitutions).clone();

        ctx.instantiate_nodes(db, solver, &mut nodes);
        ctx.instantiate_parameters(db, solver, &mut parameters);

        let substitutions = solver.insert_substitutions(nodes, parameters);

        Some(Box::new(InstantiateConstraint {
            source_node: ctx.source_node,
            bound_path: ctx.bound_path.clone(),
            definition: self.definition,
            substitutions,
            traces: ctx.instantiate_traces(db, solver, &self.traces),
        }))
    }

    fn run(self: Box<Self>, db: &mut Db, solver: &mut Solver) -> RunResult {
        // NOTE: Types are *not* applied before instantiating; we have access to
        // all related nodes/constraints here, which together will form better
        // groups

        let DefinitionConstraints(constraints) =
            db.get(self.definition).cloned().unwrap_or_default();

        let mut ctx = InstantiateCtx {
            definition: self.definition,
            source_node: self.source_node,
            bound_path: self.bound_path.clone(),
            substitutions: self.substitutions,
        };

        let mut instantiated_constraints = constraints
            .into_iter()
            .flat_map(|constraint| constraint.instantiate(db, solver, &mut ctx))
            .collect::<Vec<_>>();

        let mut traces = self.traces.clone();
        for trace in &mut traces {
            trace.from.extend(solver.active_traces.clone());
        }

        for constraint in &mut instantiated_constraints {
            constraint.traces_mut().extend(traces.iter().cloned());
        }

        RunResult::Insert(instantiated_constraints)
    }
}
